crates/red_knot_python_semantic/resources/mdtest/mro.md

```diff
@@ -0,0 +1,377 @@
+# Method Resolution Order tests
+
+Tests that assert that we can infer the correct type for a class's
+`__mro__` attribute.
+
+This attribute is rarely accessed directly at runtime. However, it's
+extremely important for *us* to know the precise possible values of
+a class's Method Resolution Order, or we won't be able to infer the
+correct type of attributes accessed from instances.
+
+For documentation on method resolution orders, see:
+
+- <https://docs.python.org/3/glossary.html#term-method-resolution-order>
+- <https://docs.python.org/3/howto/mro.html#python-2-3-mro>
+
+## No bases
+
+```py
+class C:
+    pass
+
+reveal_type(C.__mro__)  # revealed: tuple[Literal[C], Literal[object]]
+```
+
+## The special case: `object` itself
+
+```py
+reveal_type(object.__mro__)  # revealed: tuple[Literal[object]]
+```
+
+## Explicit inheritance from `object`
+
+```py
+class C(object):
+    pass
+
+reveal_type(C.__mro__)  # revealed: tuple[Literal[C], Literal[object]]
+```
+
+## Explicit inheritance from non-`object` single base
+
+```py
+class A:
+    pass
+
+class B(A):
+    pass
+
+reveal_type(B.__mro__)  # revealed: tuple[Literal[B], Literal[A], Literal[object]]
+```
+
+## Linearization of multiple bases
+
+```py
+class A:
+    pass
+
+class B:
+    pass
+
+class C(A, B):
+    pass
+
+reveal_type(C.__mro__)  # revealed: tuple[Literal[C], Literal[A], Literal[B], Literal[object]]
+```
+
+## Complex diamond inheritance (1)
+
+This is "ex_2" from <https://docs.python.org/3/howto/mro.html#the-end>
+
+```py
+class O:
```

Comment:
I'm not a massive fan of using `O` as a variable name here, because it's so easy to misread it as `0`. However, I experimented with using a different variable name, and I found it even more confusing, because it was very hard to remember which class it corresponded to in the `ex_2` example in https://docs.python.org/3/howto/mro.html#the-end. Unfortunately that article uses `O` as the class name.